rfc/0011-value-transformer/README.md

```diff
@@ -0,0 +1,191 @@
+
+SPDX-License-Identifier: AGPL-3.0-only
+-->
+
+# RFC 0000: Value Transformer
+
+| | |
+|---|---|
+| Feature Tag | `value-transformer` |
+| Status | `DRAFT` | <!-- Possible values: DRAFT, DISCUSSION, ACCEPTED, REJECTED -->
+| Responsible | `georg-schwarz` |
+<!-- 
+  Status Overview:
+  - DRAFT: The RFC is not ready for a review and currently under change. Feel free to already ask for feedback on the structure and contents at this stage.
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+
+This RFC introduces the concept of Value Transformers that allow reading values, writing values, and transforming values.
+
+**Note:** This RFC focuses enhancing the current state but has an elaborate "Possible Enhancements" section that showcases how this concept can be extended in the future.
+
+## Motivation
+
+Right now, we can only "parse" values but not manipulate them in order to clean data.
+
+## Explanation
+
+```
+transformer <name> {
+  input <inputName> oftype <valuetype>; // potentially many
+  output <outputName> oftype <valuetype>; // for now: exactly one
+
+  <outputName>: <expression producing output valuetype>;
+}
+```
+
+A transformer can map none, one, or multiple inputs to at least one output with their types.
+See future enhancements for multiple outputs.
+
+The output value is assigned (`:`) to an expression to produces the output values.
+
+### Example 1: Simple numeric transformation
+```
+transformer CelsiusToKelvin {
+  input tempCelsius oftype decimal;
+  output tempKelvin oftype decimal;
+
+  tempKelvin: tempCelsius + 273.15; // simple calculation
+}
+```
+
+### Example 2: Multiple inputs
+```
+// "MyStreet" and "7b" => "MyStreet 7b"
+transformer AddressComposer {
+  input streetName oftype text;
+  input houseNumber oftype text;
+  output address oftype text;
+
+  address: streetName + " " + houseNumber; // assumption: "+" is string appending (separate RFC)
+}
+```
+
+
+### Example 3: Multiple outputs
+```
+// 456 => (ones:6, tens:5, hundreds:4)
+transformer DigitSeparator {
+  input number oftype integer;
+  output ones oftype integer;
+  output tens oftype integer;
+  output hundreds oftype integer;
+
+  ones: number % 10;
+  tens: floor (number / 10) % 10;
+  hundreds: floor (number / 100);
+}
+```
+
+### Usage 1: TableInterpreter
+
+The `TableInterpreter` block uses a transformer from `text` to the desired value type with the `with` keyword. Parsing can still be used as currently (`as` keyword)
+
+Example:
+```
+block GasReserveTableInterpreter oftype TableInterpreter {
+  header: true;
+  columns: [ // only allow transformers with one input and one output or plain parsing here
+      "Datum" with DateTransformer,
```

Comment:
I don't feel it is a good idea to bring transformers into this, we should focus on only using valuetypes consistently.

So I think, this line should always be:

`"Datum" as date`

For defining a transformer, I think it could be okay to apply it here with `with` or `using`:

`"Datum" as date using DateTransformer`

If possible, we could try to find a chain of transformers that automatically work for the input type to the output type. 

The core of my objection here is not using valuetypes consistently though.

Reply:
In general, I thing we could replace valuetype statements with valuetypes and transformers? Consider:

- A valuetype A
- A valuetype B
- A transformer from A to B

`<valuetypeA>` is currently a valid valuetype expression. We could replace it with `<valuetypeB> using <transformerAtoB>`. That would lead to exactly my suggested syntax and could be used in any place we currently use valuetypes. It also automatically allows chaining: `<valuetypeC> using <transformerBtoC> using <transformerAtoB>`.

Reply:
Let me rephrase to check if I understood correctly:

We introduce a concept `ValueTypeExpression` that always evaluates to a `ValueType`. The syntax is as follows:
```
<final output valuetype> using <transformer from input to output type> (using <transformer from input to output type>)*
```
We would introduce a set of default transormers that allow parsing from text to the primitive ValueType.

Reply:
I'd start with using and move to nicer syntax later.

And I agree on the `ValueTypeExpression`. Unsure what you mean by default transformers, if you refer to built-in valuetypes I'd just have them auto parse as now. Primitive valuetypes are user defined so I am unsure how we would introduce default transformers for them?